### PR TITLE
Project: Private Events

### PR DESCRIPTION
**Pull request by crappypapa**
In this project, we used ActiveRecord’s associations to model the workings of a private event website.
We created associations between our models so as to satisfy the following conditions

- A user can create events. 
- A user can attend many events. 
- An event can be attended by many users.

We have done this project with little regard for styling as we would like to be certain the app meets requirements before expending energy on styling.

### CODE REVIEW

**Comment by mimipeshy** [comment]
## STATUS: CHANGES REQUESTED 🔴 ♻️ 

Hello team @crappypapa and @CollinsTatang 👋🏾 👋🏾  great job on working on this milestone

### Highlights

- You have followed the correct git-flow ✔️ 
- Your linter checks do not raise any error ✔️ 
-  A user can create events and attend events ✔️ 

### To Improve

- [x] In your README file, a user will need to migrate the database before running the command to open the server. Kindly add it
- [x] Since this is not a console project, the user should run `rails s` to start the rails server. Kindly fix that as well
- [x] Your MIT and issues links are broken as well.
- [x] Check inline comments on what can be improved 😄 


That's all. 😄 , kindy submit another review when you are done

Regards
Peris Ndanu
_If you have any questions, kindly leave a comment here and tag my name @mimipeshy and I will be glad to help_

File: app/controllers/events_controller.rb [hunk 1]
```diff
@@ -0,0 +1,86 @@
+class EventsController < ApplicationController
+  before_action :set_event, only: %i[show edit update destroy]
+  before_action :require_login, except: [:index]
+
+  # GET /events or /events.json
+  def index
+    @events = Event.all
+  end
+
```
Comment (mimipeshy) [inline app/controllers/events_controller.rb]: - [x]  Event #index should list past events, i.e. `@past_events = Event.previous_events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from  [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

- [x] Event #index should list upcoming events, i.e. `@upcoming_events = Event.upcoming_events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  end
+
+  def new
+    @user = User.new
+  end
+
+  def show
+    @user = User.find(params[:id])
+  end
+
```
Comment (mimipeshy) [inline app/controllers/users_controller.rb]: ##### UsersController

###### `show` action

* [x] lists events created by user, i.e. `@created_events = current_user.events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/current_user_show_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
* [x] lists events previoulsy attendented by user **using event model scope**, i.e. `@past_events = current_user.attended_events.previous` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
* [x] lists upcoming events previously attended by user **using event model scope**, i.e. `@upcoming_events = current_user.attended_events.upcoming` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

File: config/routes.rb [hunk 3]
```diff
@@ -1,3 +1,19 @@
...
+  root 'events#index'
+  get 'signup' => 'users#new', as: 'signup'
+  get 'login' => 'sessions#new', as: 'login'
+  get 'logout' => 'sessions#destroy', as: 'logout'
+  resources :events do
+    member do
+      get 'attend'
+      get 'cancel_attend'
+    end
+  end
```
Comment (mimipeshy) [inline config/routes.rb]: \[**OPTIONAL**\] 🟢 
- [x] Lets also ensure that  routes are only for used actions, i.e. `resources :events, only: [:new, :create, :show, :index]`

File: app/views/events/_form.html.erb [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    <div id="error_explanation">
+      <h2><%= pluralize(event.errors.count, "error") %> prohibited this event from being saved:</h2>
+
+      <ul>
+        <% event.errors.each do |error| %>
+          <li><%= error.full_message %></li>
+        <% end %>
+      </ul>
+    </div>
+  <% end %>
```
Comment (rukundoeric) [inline app/views/events/_form.html.erb]: Move this logic to helpers

File: app/views/events/show.html.erb [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+<div >
+  <li >This event has no attendees yet</li>
+</div>
+<% else %>
+  <div>
+    <% @event.attendees.each do |attendee| %>
+    <li> <%= attendee.name %> </li>
+    <% end %>
+  </div>
+<% end %>
```
Comment (rukundoeric) [inline app/views/events/show.html.erb]: Move this logic to helpers

File: app/views/events/show.html.erb [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+  
+    <% if current_user == @event.creator %>
+     
+      <%= link_to 'Delete', event_path, method: :delete, class: "btn",
+                            data: { confirm: "Delete this event?"} %>
+    <% elsif @event.attendees.include?(current_user) %>
+      <%= link_to 'Cancel', cancel_attend_event_path, class: "btn" %>
+    <% else %>
+      <%= link_to 'Attend', attend_event_path, class: "btn" unless @event.date < Time.now %>
+    <% end %>
```
Comment (rukundoeric) [inline app/views/events/show.html.erb]: Move this logic to helpers

File: app/views/layouts/application.html.erb [hunk 7]
```diff
@@ -11,6 +11,52 @@
   </head>
 
   <body>
+    <% if flash[:notice] %>
+      <div class="notification">
```
Comment (rukundoeric) [inline app/views/layouts/application.html.erb]: Move this logic to helpers

File: app/views/layouts/application.html.erb [hunk 8]
```diff
@@ -11,6 +11,52 @@
...
+        <p class="notice"><%= notice %></p>
+      </div>
+    <% end %>
+    <% if flash[:alert] %>
+      <div class="notification">
```
Comment (rukundoeric) [inline app/views/layouts/application.html.erb]: Move this logic to helpers

File: app/views/layouts/application.html.erb [hunk 9]
```diff
@@ -11,6 +11,52 @@
...
+              <% if current_user %>
+                <p class= 'control'>
+                  <%= link_to current_user.name, user_path(current_user), class: "btn" %>
+                </p>
+                <p class= 'control'>
+                  <%= link_to "Log out", logout_path, class: "btn" %>
+                </p>
+              <% else %>
+                <p class="control">
+                  <%= link_to 'Sign In', login_path, class: "btn" %>
```
Comment (rukundoeric) [inline app/views/layouts/application.html.erb]: Move this logic to helpers

File: app/views/users/_form.html.erb [hunk 10]
```diff
@@ -0,0 +1,23 @@
+<%= form_with(model: user) do |form| %>
+    <% if user.errors.any? %>
+      <div id="error_explanation">
+        <h2><%= pluralize(user.errors.count, "error") %> prohibited this user from being saved:</h2>
```
Comment (rukundoeric) [inline app/views/users/_form.html.erb]: Move this logic to helpers

File: config/routes.rb [hunk 11]
```diff
@@ -1,3 +1,19 @@
 Rails.application.routes.draw do
+  root 'events#index'
+  get 'signup' => 'users#new', as: 'signup'
+  get 'login' => 'sessions#new', as: 'login'
+  get 'logout' => 'sessions#destroy', as: 'logout'
```
Comment (rukundoeric) [inline config/routes.rb]: - In RESTFUL APIs we don't need to overly complicate/customize route names:
     - So user signup is simply included here:  `resources :users`
     - login in is simply: `resources :sessions, only: [:create]`
     - And logout is simply: `resources :sessions, only: [:create]`

**Review by rukundoeric — CHANGES_REQUESTED** [review]
# Status: ♻️ CHANGES REQUIRED ♻️
_Great job guys so far 👏 💯 👍. However, there are few issues we need to address before we move on, please check the comments below._

- ### To Highlight
  - The codebasee is clean. 💯
  - The project follows Git flow. 💯
  - All feedback from the previous review are well implemented. 💯 

- ### To Improve
  - [x] [MANDATORY] Views should be logic-free. kindly remove all logic related code from views, especialy `if-else` statements.
  - [x] [MANDATORY]In RESTFUL APIs we don't need to overly complicate/customize route names:
       - So user signup is simply included here: `resources :users`
       - login in is simply: `resources :sessions, only: [:create]`
       - And logout is simply: `resources :sessions, only: [:create]`


_Please make all the required changes before you submit another code review. If you have any problems or concerns, you can reach out to me on Slack: Rukundo Eric 🚀 . I wish you all the very best 🥇_

### _Keep working hard._
### _Happy Coding😃_
-----------------------------------------------------------------------------------------------------------------------------------------
**Regards,
[Rukundo Eric](https://rukundo-eric.herokuapp.com/) ,
[LinkedIn](https://www.linkedin.com/in/rukundo-eric/), 
Slack: Rukundo Eric**

**Comment by crappypapa** [comment]
I have removed all logic from views as requested and I have refactored the session routes to the best of my current abilities.

**Review by Dipeshtwis — APPROVED** [review]
# Status: Approved ✔️

## Hi @crappypapa &  @CollinsTatang 

Good job on implementing the suggested changes. Well done. 

- In readme file setup section, you have missed the `rails db:migrate` command to populate the database. Please write this step before merging.

**You may close this Pull Request and merge the changes**
 
**Keep being awesome devs and good luck with your next milestone!**
---------

Cheers and Happy coding!👏👏👏

Code review by **Dipesh Kumar**
(Feel free to contact me on slack by the handle of  **deep** if you have any additional questions).
[Dipesh Kumar](https://deeptwist.tk) | [Twitter](https://twitter.com/97deepeshkumar)

**Comment by crappypapa** [comment]
@Dipeshtwis  I will add the required command to the readme as suggested. Thank you.

**Pull request merged**